File: backend/tests/factories.py
"""Shared test data factories — small async builder functions that create
and flush a fully-shaped row with sane defaults, overridable per call.
Used instead of every test file hand-rolling its own `_create_repository`-
style helper (several already did, independently, before this file
existed): a test that only cares about a repository's `language` shouldn't
have to also restate its organization, installation, and every other
required field, and when a model gains a required column, only this file
needs to change.

Every function flushes (never commits) so the caller gets a real
generated id/timestamps back immediately, while still controlling its own
transaction boundary — matches the pattern already used throughout
tests/integration/."""

import uuid
from datetime import UTC, datetime

from sqlalchemy.ext.asyncio import AsyncSession

from app.models.conversation import Conversation
from app.models.organization import Organization
from app.models.organization_member import OrganizationMember
from app.models.pull_request import PullRequest
from app.models.repository import Repository
from app.models.repository_membership import RepositoryMembership
from app.models.role import Role
from app.models.user import User
from app.repositories import (
    conversation_repository,
    github_installation_repository,
    organization_member_repository,
    organization_repository,
    pull_request_repository,
    repository_membership_repository,
    repository_repository,
    user_repository,
)


def _unique(prefix: str) -> str:
    return f"{prefix}-{uuid.uuid4().hex[:8]}"


async def create_user(
    db: AsyncSession,
    *,
    email: str | None = None,
    full_name: str = "Test User",
    password_hash: str | None = None,
) -> User:
    user = user_repository.create(
        db,
        email=email or f"{_unique('user')}@example.com",
        full_name=full_name,
        password_hash=password_hash,
    )
    await db.flush()
    return user


async def create_organization(
    db: AsyncSession, *, name: str = "Acme Inc", slug: str | None = None
) -> Organization:
    organization = organization_repository.create(db, name=name, slug=slug or _unique("acme"))
    await db.flush()
    return organization


async def create_organization_member(
    db: AsyncSession,
    *,
    organization: Organization | None = None,
    user: User | None = None,
    role: Role = Role.OWNER,
) -> OrganizationMember:
    organization = organization or await create_organization(db)
    user = user or await create_user(db)
    member = organization_member_repository.create(
        db, organization_id=organization.id, user_id=user.id, role=role
    )
    await db.flush()
    return member


async def create_repository(
    db: AsyncSession,
    *,
    organization: Organization | None = None,
    full_name: str = "acme/widgets",
    name: str = "widgets",
    language: str | None = "Python",
    private: bool = False,
) -> Repository:
    organization = organization or await create_organization(db)
    installation = github_installation_repository.create(
        db,
        organization_id=organization.id,
        github_installation_id=uuid.uuid4().int % 1_000_000_000,
        account_login=organization.slug,
        account_type="Organization",
    )
    await db.flush()
    repository = repository_repository.create(
        db,
        organization_id=organization.id,
        installation_id=installation.id,
        github_repo_id=uuid.uuid4().int % 1_000_000_000,
        full_name=full_name,
        name=name,
        description=None,
        language=language,
        stargazers_count=0,
        forks_count=0,
        default_branch="main",
        private=private,
        html_url=f"https://github.com/{full_name}",
    )
    await db.flush()
    return repository


async def create_repository_membership(
    db: AsyncSession, *, repository: Repository, user: User
) -> RepositoryMembership:
    membership = repository_membership_repository.create(
        db, repository_id=repository.id, user_id=user.id
    )
    await db.flush()
    return membership


async def create_pull_request(
    db: AsyncSession,
    *,
    repository: Repository,
    number: int = 1,
    title: str = "Add a feature",
    state: str = "open",
    author_login: str | None = "octocat",
    head_sha: str | None = None,
    merged_at: datetime | None = None,
    closed_at: datetime | None = None,
) -> PullRequest:
    now = datetime.now(UTC)
    pr = pull_request_repository.create(
        db,
        repository_id=repository.id,
        number=number,
        title=title,
        state=state,
        author_login=author_login,
        html_url=f"{repository.html_url}/pull/{number}",
        head_sha=head_sha or ("a" * 40),
        github_created_at=now,
        github_updated_at=now,
        closed_at=closed_at,
        merged_at=merged_at,
        synced_at=now,
    )
    await db.flush()
    return pr


async def create_conversation(
    db: AsyncSession, *, repository: Repository, created_by: User
) -> Conversation:
    conversation = conversation_repository.create(
        db, repository_id=repository.id, created_by_user_id=created_by.id
    )
    await db.flush()
    return conversation

File: frontend/test/factories.ts
/**
 * Test data factories — small builder functions that return a fully-shaped
 * object with sane defaults, overridable per call. Used instead of copying
 * a full literal into every test: a test that only cares about `role`
 * shouldn't have to restate every other field of a Member, and when the
 * `Member` shape changes, only this file needs to change, not every test
 * that builds one. Keep these in sync with lib/types.ts.
 */
import type {
  Commit,
  Conversation,
  IndexingJob,
  Member,
  Organization,
  OrganizationMembership,
  PullRequest,
  PullRequestAnalysis,
  Repository,
  SourceReference,
  User,
} from "@/lib/types";

let counter = 0;
function nextId(prefix: string): string {
  counter += 1;
  return `${prefix}-${counter}`;
}

export function makeUser(overrides: Partial<User> = {}): User {
  const id = overrides.id ?? nextId("user");
  return {
    id,
    email: `${id}@example.com`,
    full_name: "Test User",
    avatar_url: null,
    email_verified: true,
    created_at: "2026-01-01T00:00:00Z",
    ...overrides,
  };
}

export function makeOrganization(overrides: Partial<Organization> = {}): Organization {
  const id = overrides.id ?? nextId("org");
  return {
    id,
    name: "Acme Inc",
    slug: "acme-inc",
    plan: "free",
    created_at: "2026-01-01T00:00:00Z",
    ...overrides,
  };
}

export function makeOrganizationMembership(
  overrides: Partial<OrganizationMembership> = {},
): OrganizationMembership {
  return {
    organization: makeOrganization(),
    role: "owner",
    ...overrides,
  };
}

export function makeMember(overrides: Partial<Member> = {}): Member {
  return {
    user: makeUser(),
    role: "developer",
    created_at: "2026-01-01T00:00:00Z",
    ...overrides,
  };
}

export function makeRepository(overrides: Partial<Repository> = {}): Repository {
  const id = overrides.id ?? nextId("repo");
  return {
    id,
    full_name: "acme/widgets",
    name: "widgets",
    description: "A widget factory",
    language: "TypeScript",
    stargazers_count: 12,
    forks_count: 2,
    default_branch: "main",
    private: false,
    html_url: "https://github.com/acme/widgets",
    status: "ready",
    sync_error: null,
    last_synced_at: "2026-01-01T00:00:00Z",
    connected_at: "2026-01-01T00:00:00Z",
    ...overrides,
  };
}

export function makePullRequest(overrides: Partial<PullRequest> = {}): PullRequest {
  return {
    number: 1,
    title: "Add feature",
    state: "open",
    author_login: "octocat",
    html_url: "https://github.com/acme/widgets/pull/1",
    head_sha: "a".repeat(40),
    github_created_at: "2026-01-01T00:00:00Z",
    github_updated_at: "2026-01-01T00:00:00Z",
    closed_at: null,
    merged_at: null,
    ...overrides,
  };
}

export function makePullRequestAnalysis(
  overrides: Partial<PullRequestAnalysis> = {},
): PullRequestAnalysis {
  return {
    id: nextId("pr-analysis"),
    head_sha: "a".repeat(40),
    status: "succeeded",
    risk_level: "low",
    summary: "Small, well-tested change.",
    affected_components: [],
    potential_concerns: [],
    recommended_tests: [],
    files_analyzed: [],
    model: "claude-opus-5",
    error: null,
    created_at: "2026-01-01T00:00:00Z",
    started_at: "2026-01-01T00:00:00Z",
    finished_at: "2026-01-01T00:00:05Z",
    ...overrides,
  };
}

export function makeIndexingJob(overrides: Partial<IndexingJob> = {}): IndexingJob {
  const id = overrides.id ?? nextId("job");
  return {
    id,
    repository_id: overrides.repository_id ?? nextId("repo"),
    trigger: "manual",
    commit_sha: "a".repeat(40),
    status: "succeeded",
    current_stage: "done",
    files_discovered: 100,
    files_processed: 100,
    files_skipped: 0,
    symbols_extracted: 500,
    chunks_created: 500,
    embeddings_generated: 500,
    started_at: "2026-01-01T00:00:00Z",
    finished_at: "2026-01-01T00:00:10Z",
    error: null,
    created_at: "2026-01-01T00:00:00Z",
    errors: [],
    ...overrides,
  };
}

export function makeSourceReference(overrides: Partial<SourceReference> = {}): SourceReference {
  return {
    source_type: "vector",
    rank: 1,
    score: 0.92,
    file_path: "src/auth/login.ts",
    start_line: 10,
    end_line: 20,
    symbol_name: "login",
    commit_sha: null,
    commit_url: null,
    ...overrides,
  };
}

export function makeCommit(overrides: Partial<Commit> = {}): Commit {
  return {
    sha: "a".repeat(40),
    message: "Fix bug",
    author_name: "Octocat",
    author_login: "octocat",
    html_url: "https://github.com/acme/widgets/commit/" + "a".repeat(40),
    authored_at: "2026-01-01T00:00:00Z",
    ...overrides,
  };
}

export function makeConversation(overrides: Partial<Conversation> = {}): Conversation {
  return {
    id: nextId("conversation"),
    repository_id: overrides.repository_id ?? nextId("repo"),
    title: "How does auth work?",
    created_at: "2026-01-01T00:00:00Z",
    updated_at: "2026-01-01T00:00:00Z",
    ...overrides,
  };
}
